File: 백준/Silver/2178. 미로 탐색/미로 탐색.java
import java.util.*;
import java.io.*;

public class Main{

    static int[] dx = {-1,1,0,0};
    static int[] dy = {0,0,-1,1};
    static Queue<int[]> queue;

    static int N,M;

    static int[][] graph;
    static boolean[][] visited;

    public static void main(String[] args)throws Exception{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        N = Integer.parseInt(st.nextToken());
        M = Integer.parseInt(st.nextToken());
        graph = new int[N][M];
        visited = new boolean[N][M];
        for(int i=0; i<N; i++){
            String s = br.readLine();
            for(int j=0; j<M; j++){
                graph[i][j] = s.charAt(j) - '0';
            }
        }
        bfs(0,0);
        System.out.println(graph[N-1][M-1]);
    }
    static void bfs(int x, int y){
        queue = new LinkedList<>();
        queue.offer(new int[]{x,y});
        visited[x][y] = true;
        while(!queue.isEmpty()){
            int[] poll = queue.poll();
            int curX = poll[0];
            int curY = poll[1];
            for(int i=0; i<4; i++){
                int nx = curX + dx[i];
                int ny = curY + dy[i];
                if(nx < 0 || nx >= N || ny <0 || ny>= M){
                    continue;
                }
                if(!visited[nx][ny] && graph[nx][ny] == 1){
                    graph[nx][ny] = graph[curX][curY] + 1;
                    queue.offer(new int[]{nx,ny});
                    visited[nx][ny] = true;
                }
            }
        }
    }
}

File: 백준/Silver/2178. 미로 탐색/미로 탐색.js
const readline = require("readline");

const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
});

let inputLine = 0;
let N = 0;
let M = 0;
let arr = [];
let visited;
let dx = [-1, 1, 0, 0];
let dy = [0, 0, -1, 1];
rl.on("line", (line) => {
  if (inputLine == 0) {
    let userInput = line.split(" ").map((el) => parseInt(el));
    N = userInput[0];
    M = userInput[1];
    visited = Array.from({ length: N }, () =>
      Array.from({ length: M }, () => false)
    );
  }
  if (inputLine > 0 && inputLine <= N) {
    arr.push(line.split("").map((el) => parseInt(el)));
  }
  if (inputLine == N) {
    solution();
    process.exit();
  }
  inputLine++;
});

function solution() {
  bfs(0, 0);
  console.log(arr[N - 1][M - 1]);
  process.exit();
}

function bfs(x, y) {
  let queue = [];
  queue.push([x, y]);
  visited[x][y] = true;
  while (!queue.length == 0) {
    let [curX, curY] = queue.shift();
    for (let i = 0; i < 4; i++) {
      let nx = curX + dx[i];
      let ny = curY + dy[i];
      if (nx < 0 || ny < 0 || nx >= N || ny >= M) {
        continue;
      }
      if (!visited[nx][ny] && arr[nx][ny] == 1) {
        queue.push([nx, ny]);
        arr[nx][ny] = arr[curX][curY] + 1;
      }
    }
  }
}
